### ISSUE TITLE
ACCESS_VIOLATION in TestArrays.testAdd and testWarmUp with several OpenCL devices

### ISSUE THREAD

**Issue by vsilaev** [beehive-lab/TornadoVM#56]
In my configuration I have 3 OpenCL devices (listed in the same order as in Tornado)
1. GeForce CTX 1080Ti (GPU)
2. AMD FirePro W5000 (GPU)
3. Intel i7 4930K (CPU)

I'm trying to run test suite on AMD device. I see no option to re-order devices or specify a default device, so I set `tornado` args as following:
```
TORNADO_FLAGS="-Djava.library.path=${TORNADO_SDK}/lib -Ds0.t0.device=0:1 -Dtornado.opencl.compiler.options=\"-w\""
```

When running tests mentioned is subject, I've got ACCESS_VIOLATION, because Tornado switched between devices when scheduling additional tasks:
```
$ tornado -Dtornado.unittests.verbose=True -Xmx6g -Dtornado.recover.bailout=False  --debug uk.ac.manchester.tornado.unittests.tools.TornadoTestRunner  uk.ac.manchester.tornado.unittests.arrays.TestArrays
task info: s0.t0
        platform          : AMD Accelerated Parallel Processing
        device            : Pitcairn CL_DEVICE_TYPE_GPU (available)
        dims              : 1
        global work offset: [0]
        global work size  : [4096]
        local  work size  : [1024, 1, 1]

task info: s0.t0
        platform          : AMD Accelerated Parallel Processing
        device            : Pitcairn CL_DEVICE_TYPE_GPU (available)
        dims              : 1
        global work offset: [0]
        global work size  : [128]
        local  work size  : [128, 1, 1]

[TornadoVM-OCL-JNI] ERROR : clEnqueueWriteBuffer -> Returned: -58
task info: s0.t1
        platform          : NVIDIA CUDA
        device            : GeForce GTX 1080 Ti CL_DEVICE_TYPE_GPU (available)
        dims              : 1
        global work offset: [0]
        global work size  : [128]
        local  work size  : [128, 1, 1]

#
# A fatal error has been detected by the Java Runtime Environment:
#
#  EXCEPTION_ACCESS_VIOLATION (0xc0000005) at pc=0x00007ffb4d6a00ef, pid=23784, tid=0x0000000000005420
#
# JRE version: OpenJDK Runtime Environment (8.0_262-b10) (build 1.8.0_262-b10)
# Java VM: OpenJDK 64-Bit Server VM GraalVM CE 20.2.0 (25.262-b10-jvmci-20.2-b03 mixed mode windows-amd64 )
# Problematic frame:
# C  [nvopencl64.dll+0x200ef]
#
# Failed to write core dump. Minidumps are not enabled by default on client versions of Windows
#
# An error report file with more information is saved as:
# D:\Projects\TornadoVM\hs_err_pid23784.log
#
# If you would like to submit a bug report, please visit:
#   https://github.com/oracle/graal/issues
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#

```

See how AMD switched to NVIDIA for s0:t1 after s0:t0. And error -58 means just this -- wrong event, coming from another device.

Probably there is an undocumented option to specify default device, but I guess it's wrong to change device in this case: data is streamed-in on another device, so why ever another device should be selected?

**Comment by stratika** [comment]
Hi @vsilaev, I am not sure what is that you would like to do. From what I see, you have used the flag: `-Ds0.t0.device=0:1` which sets the task `t0` from TaskSchedule `s0` to be executed on the device 0:1 (AMD GPU). On the other hand, the task `t1` can be executed on any device, since there is no flag driving where to execute (e.g. `-Ds0.t1.device=0:1`). So, it is scheduled on the default (the first which is the Nvidia GPU).

Can you be more descriptive about the bug?

**Comment by vsilaev** [comment]
Ok, why subsequent tasks doesn't inherit original task's device context?

A. It's logical
B. How it's ever possible to use another device if input (stream-in, copy-in) was put to the original one? And events are fired from the original one?

**Comment by jjfumero** [comment]
I can't reproduce this error in our AMD (which uses OpenCL 2.0). 

We recently discovered an issue with the AMD and events. It seems there are some events that are passed wrongly. 

Could you use this export and check again?

`export AMD_OCL_WAIT_COMMAND=1` 

Regarding device-switching. Unless the user sets a new device either from the command line, or an API call, TornadoVM will not change from the default device. Also, if the dynamic reconfiguration is used, then TornadoVM will change the device during runtime. But these tests do not use dynamic reconfiguration.

**Comment by vsilaev** [comment]
> We recently discovered an issue with the AMD and events. It seems there are some events that are passed wrongly.

Yes, I've fixed this along with adding async support. Fixed this already and will share once I'll be able to test on AMD. NVIDIA has 325 out of 330 tests passed (some profiler test is broken in async), but in general NVIDIA is not the best candidate to be async due to [synchronous clEnqueueNDRangeKernel behavior](https://www.google.com/search?q=clenqueuendrangekernel+nvidia+blocking).
